### PR TITLE
logging: Anomaly logging 

### PR DESCRIPTION
**Pull request by jlucovsky**
This changeset adds anomaly logging to suricata for [issue 2282](https://redmine.openinfosecfoundation.org/issues/2282)

Anomaly logging is controlled via the anomaly section within eve-log.
There is a single option -- packethdr -- for including the packet header
in the anomaly.

Note: continuation of #3778

Example output

Packet hdr logging off:

```
{"timestamp":"1969-12-31T16:00:00.000000-0800","flow_id":196670847451136,"pcap_cnt":3,"event_type":"anomaly","src_ip":"204.252.102.11","dest_ip":"206.65.98.18","proto":"ICMP","icmp_type":24,"icmp_code":79,"anomaly":{"event":"decoder.icmpv4.unknown_type"}}
{"timestamp":"1969-12-31T16:00:00.000000-0800","pcap_cnt":4,"event_type":"anomaly","src_ip":"208.21.2.184","src_port":0,"dest_ip":"10.1.1.99","dest_port":0,"proto":"UDP","anomaly":{"event":"decoder.udp.pkt_too_small"}}
{"timestamp":"1969-12-31T16:00:00.000000-0800","anomaly":{"decoder-event":"decoder.ipv4.wrong_ip_version"}}
```

Packet hdr logging enabled:
```
{"timestamp":"1969-12-31T16:00:00.000000-0800","flow_id":79166984683520,"pcap_cnt":3,"event_type":"anomaly","src_ip":"204.252.102.11","dest_ip":"206.65.98.18","proto":"ICMP","icmp_type":24,"icmp_code":79,"packethdr":"FF FF FF FF FF FF 01 01 01 01 01 01 08 00 45 00 00 54 8F B3 40 00 FD 01 8A 99 CC FC 66 0B CE 41 ","anomaly":{"event":"decoder.icmpv4.unknown_type"}}
{"timestamp":"1969-12-31T16:00:00.000000-0800","pcap_cnt":4,"event_type":"anomaly","src_ip":"208.21.2.184","src_port":0,"dest_ip":"10.1.1.99","dest_port":0,"proto":"UDP","packethdr":"FF FF FF FF FF FF 01 01 01 01 01 01 08 00 45 00 00 3C C5 9E 40 00 FF 11 D7 E0 D0 15 02 B8 0A 01 ","anomaly":{"event":"decoder.udp.pkt_too_small"}}
{"timestamp":"1969-12-31T16:00:00.000000-0800","packethdr":"FF FF FF FF FF FF 01 01 01 01 01 01 08 00 E4 20 A0 CC 6E F0 1E B7 C0 DF 1A 6B FA 6D 57 28 25 73 ","anomaly":{"decoder-event":"decoder.ipv4.wrong_ip_version"}}
```

### CODE REVIEW

File: src/output-json-anomaly.c [hunk 1]
```diff
@@ -0,0 +1,354 @@
...
+    AnomalyJsonOutputCtx* json_output_ctx;
+} JsonAnomalyLogThread;
+
+static int AnomalyJson(ThreadVars *tv, JsonAnomalyLogThread *aft, const Packet *p)
+{
+    if (p->events.cnt == 0) {
+        return TM_ECODE_OK;
+    }
+
+    bool isIPpkt = PKT_IS_IPV4(p) || PKT_IS_IPV6(p);
```
[the remark this answers is not in the document]
Comment (jlucovsky) [reply src/output-json-anomaly.c]: Will fix.

File: src/output-json-anomaly.c [hunk 2]
```diff
@@ -0,0 +1,354 @@
...
+            return TM_ECODE_OK;
+        }
+
+        json_t *ajs = json_object();
+        if (unlikely(ajs == NULL)) {
+            json_decref(js);
+            return TM_ECODE_OK;
+        }
+
+        if (p->tenant_id > 0)
```
[the remark this answers is not in the document]
Comment (jlucovsky) [reply src/output-json-anomaly.c]: Will remove.

File: src/output-json-anomaly.c [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+
+        uint8_t event_code = p->events.events[i];
+        const char *event;
+        if (EVENT_IS_DECODER_PACKET_ERROR(event_code)) {
+            event = DEvents[event_code].event_name;
+        } else {
+            event = (const char *) "Unknown";
+        }
+        json_object_set_new(ajs,
+                            isIPpkt ? "event" : "decoder-event",
```
[the remark this answers is not in the document]
Comment (jlucovsky) [reply src/output-json-anomaly.c]: Will use `event` then.

File: src/output-json-anomaly.c [hunk 4]
```diff
@@ -0,0 +1,354 @@
...
+    return p->events.cnt > 0 ? TRUE : FALSE;
+}
+
+#define OUTPUT_BUFFER_SIZE 65535
+static TmEcode JsonAnomalyLogThreadInit(ThreadVars *t, const void *initdata, void **data)
+{
+    JsonAnomalyLogThread *aft = SCMalloc(sizeof(JsonAnomalyLogThread));
+    if (unlikely(aft == NULL))
+        return TM_ECODE_FAILED;
+    memset(aft, 0, sizeof(JsonAnomalyLogThread));
```
[the remark this answers is not in the document]
Comment (jlucovsky) [reply src/output-json-anomaly.c]: Will modify to use `SCCalloc` instead.

File: src/output-json-anomaly.c [hunk 5]
```diff
@@ -0,0 +1,354 @@
...
+        return result;
+    }
+
+    OutputCtx *output_ctx = SCCalloc(1, sizeof(OutputCtx));
+    if (unlikely(output_ctx == NULL)) {
+        LogFileFreeCtx(logfile_ctx);
+        return result;
+    }
+
+    json_output_ctx = SCMalloc(sizeof(AnomalyJsonOutputCtx));
```
[the remark this answers is not in the document]
Comment (jlucovsky) [reply src/output-json-anomaly.c]: Will adjust.

File: src/output-json-anomaly.c [hunk 6]
```diff
@@ -0,0 +1,354 @@
...
+{
+    OutputInitResult result = { NULL, false };
+    OutputJsonCtx *ajt = parent_ctx->data;
+    AnomalyJsonOutputCtx *json_output_ctx = NULL;
+
+    OutputCtx *output_ctx = SCCalloc(1, sizeof(OutputCtx));
+    if (unlikely(output_ctx == NULL))
+        return result;
+
+    json_output_ctx = SCMalloc(sizeof(AnomalyJsonOutputCtx));
```
[the remark this answers is not in the document]
Comment (jlucovsky) [reply src/output-json-anomaly.c]: Will adjust.

File: src/output-json-anomaly.c [hunk 7]
```diff
@@ -0,0 +1,354 @@
...
+
+    result.ctx = output_ctx;
+    result.ok = true;
+    return result;
+
+error:
+    if (json_output_ctx != NULL) {
+        SCFree(json_output_ctx);
+    }
+    if (output_ctx != NULL) {
```
[the remark this answers is not in the document]
Comment (jlucovsky) [reply src/output-json-anomaly.c]: Good catch; will remove conditional.

File: src/output-json-anomaly.c [hunk 8]
```diff
@@ -0,0 +1,354 @@
...
+typedef struct JsonAnomalyLogThread_ {
+    /** LogFileCtx has the pointer to the file and a mutex to allow multithreading */
+    LogFileCtx* file_ctx;
+    MemBuffer *json_buffer;
+    AnomalyJsonOutputCtx* json_output_ctx;
+} JsonAnomalyLogThread;
+
+static int AnomalyJson(ThreadVars *tv, JsonAnomalyLogThread *aft, const Packet *p)
+{
+    if (p->events.cnt == 0) {
```
[the remark this answers is not in the document]
Comment (jlucovsky) [reply src/output-json-anomaly.c]: No. I modeled this after the pattern use in output-json-alert.c -- should we update that code in a separate commit on this pr?

File: src/output-json-anomaly.c [hunk 9]
```diff
@@ -0,0 +1,354 @@
...
+        } else {
+            JsonFiveTuple((const Packet *)p, LOG_DIR_PACKET, js);
+            JsonAddCommonOptions(&aft->json_output_ctx->cfg, p, p->flow, js);
+        }
+
+        if (aft->json_output_ctx->flags & LOG_JSON_PACKET) {
+            char buf[(32 * 3) + 1];
+            PrintRawLineHexBuf(buf, sizeof(buf), GET_PKT_DATA(p),
+                               GET_PKT_LEN(p) < 32 ? GET_PKT_LEN(p) : 32);
+            json_object_set_new(js, "packethdr", json_string((char *)buf));
```
[the remark this answers is not in the document]
Comment (jlucovsky) [reply src/output-json-anomaly.c]: Will add the linktype; do you want it only when the `packethdr` option is enabled?

File: src/output-json-anomaly.c [hunk 10]
```diff
@@ -0,0 +1,354 @@
...
+        }
+
+        uint8_t event_code = p->events.events[i];
+        const char *event;
+        if (EVENT_IS_DECODER_PACKET_ERROR(event_code)) {
+            event = DEvents[event_code].event_name;
+        } else {
+            event = (const char *) "Unknown";
```
[the remark this answers is not in the document]
Comment (jlucovsky) [reply src/output-json-anomaly.c]: Will do.

File: src/output-json-anomaly.c [hunk 11]
```diff
@@ -0,0 +1,354 @@
...
+            json_object_set_new(ajs, "tenant_id", json_integer(p->tenant_id));
+
+        if (!isIPpkt) {
+            json_object_set_new(js, "timestamp", json_string(timebuf));
+        } else {
+            JsonFiveTuple((const Packet *)p, LOG_DIR_PACKET, js);
+            JsonAddCommonOptions(&aft->json_output_ctx->cfg, p, p->flow, js);
+        }
+
+        if (aft->json_output_ctx->flags & LOG_JSON_PACKET) {
```
[the remark this answers is not in the document]
Comment (jlucovsky) [reply src/output-json-anomaly.c]: The suggestion is to only honor the packethdr option for decoder events, when a valid ip header doesn't exist?

File: src/output-json-anomaly.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (victorjulien) [reply src/output-json-anomaly.c]: Yes.

File: src/output-json-anomaly.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (victorjulien) [reply src/output-json-anomaly.c]: Yeah I think that makes sense, thanks!

File: src/output-json-anomaly.c (hunk 11 above)
[the remark this answers is not in the document]
Comment (victorjulien) [reply src/output-json-anomaly.c]: My idea would be that it's honored for all decoder events, but not for stream events

**Comment by jlucovsky** [comment]
Continuied in #3797

**Pull request closed without merging**